homeassistant/components/roomba/vacuum.py

```diff
@@ -294,32 +269,179 @@ async def async_update(self):
+                speed_list.append(f"{behavior}-{spray}")
+        self._speed_list = speed_list
+
+    @property
+    def supported_features(self):
+        """Flag vacuum cleaner robot features that are supported."""
+        return SUPPORT_BRAAVA
+
+    @property
+    def fan_speed(self):
+        """Return the fan speed of the vacuum cleaner."""
+        behavior, spray = self._fan_speed
+        return f"{behavior}-{spray}"
+
+    @property
+    def fan_speed_list(self):
+        """Get the list of available fan speed steps of the vacuum cleaner."""
+        return self._speed_list
+
+    async def async_set_fan_speed(self, fan_speed, **kwargs):
+        """Set fan speed."""
+        try:
+            split = fan_speed.split("-", 1)
+            behavior = split[0]
+            spray = int(split[1])
+            if behavior.capitalize() in BRAAVA_MOP_BEHAVIORS:
+                behavior = behavior.capitalize()
+            assert behavior in BRAAVA_MOP_BEHAVIORS
+            assert spray in BRAAVA_SPRAY_AMOUNT
+        except (IndexError, ValueError, AssertionError):
+            _LOGGER("No such fan speed available: %s", fan_speed)
+            return
+
+        overlap = 0
+        if behavior == MOP_STANDARD:
+            overlap = OVERLAP_STANDARD
+        elif behavior == MOP_DEEP:
+            overlap = OVERLAP_DEEP
+        else:
+            overlap = OVERLAP_EXTENDED
+        await self.hass.async_add_job(
+            self.vacuum.client.publish,
+            "delta",
+            json.dumps(
+                {
+                    "rankOverlap": overlap,
+                    "padWetness": {"disposable": spray, "reusable": spray},
+                }
+            ),
+        )
+
+    async def async_update(self):
+        """Fetch state from the device."""
+        await super().async_update()
+
+        # Get Braava state
+        state = self.vacuum_state
+        detected_pad = state.get("detectedPad")
+        mop_ready = state.get("mopReady", {})
+        lid_closed = mop_ready.get("lidClosed")
+        tank_present = mop_ready.get("tankPresent")
+        tank_level = state.get("tankLvl")
+        self._state_attrs[ATTR_DETECTED_PAD] = detected_pad
+        self._state_attrs[ATTR_LID_CLOSED] = lid_closed
+        self._state_attrs[ATTR_TANK_PRESENT] = tank_present
+        self._state_attrs[ATTR_TANK_LEVEL] = tank_level
+
+        # Mopping behavior and spray amount as fan speed
+        rank_overlap = state.get("rankOverlap", {})
+        behavior = None
+        if rank_overlap == OVERLAP_STANDARD:
+            behavior = MOP_STANDARD
+        elif rank_overlap == OVERLAP_DEEP:
+            behavior = MOP_DEEP
+        elif rank_overlap == OVERLAP_EXTENDED:
+            behavior = MOP_EXTENDED
+        pad_wetness = state.get("padWetness", {})
+        # "disposable" and "reusable" values are always the same
+        pad_wetness_value = pad_wetness.get("disposable")
+        self._fan_speed = (behavior, pad_wetness_value)
```

Comment:
This is mostly existing so no need to do this unless you feel up to it:

It looks like there isn't any I/O happening here because `self.vacuum_state` is looking at `self.vacuum.master_state` which is updated by another thread.  If thats the case, all of these can be moved directly into the properties and there can be a single `async_update` in `IRobotBase` that does nothing for now.  Ideally this would have pypi package do a callback when there is new data so it doesn't have to poll.

Reply:
I've created a [PR](https://github.com/pschmitt/roombapy/pull/32) in the pypi to add the callback and I'll finish the clean up after that is merged.